internal/oras/oras.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2024 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oras
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/file"
+	"oras.land/oras-go/v2/registry"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/credentials"
+)
+
+// Download downloads the oras artifact present at the given registry URL.
+// Usage example:
+//
+// artifact := "docker.io/company/k0s:latest"
+// fp, _ := os.CreateTemp("", "k0s-oras-artifact-*")
+// err := oras.Download(ctx, artifact, fp)
+//
+// This function expects only one artifact to be present, if none is found this
+// returns an error. The artifact is downloaded in a temporary location before
+// being copied to the target.
+func Download(ctx context.Context, url string, target io.Writer, options ...OrasOption) (err error) {
+	opts := orasOptions{}
+	for _, opt := range options {
+		opt(&opts)
+	}
+
+	creds, err := opts.auth.CredentialStore(ctx)
```

Comment:
I am not entirely sure I follow what you are saying here. On your proposal it seems like we are going to need to go through the credentials (that we are going to read from a Secret in the cluster) and filter out the authentication (user/pass) we want to use as multiple may exist for different registries. The library already provides a mechanism for selecting the right credentials so this does not seem correct.

Let me know if I misunderstood your proposal here.

Reply:
Oh, right, I didn't think about the dockerconfigjson stuff. For a single file download, this is a bit too much, but as this is needed for image pull secrets, it doesn't make sense to force users to copy around secrets just because we don't require that complexity here, of course. I do have the feeling that this functionality should live somewhere else than here. Anyhow, let's keep it here for the start. We might still move it to another location at a later stage.